Menu as GUI, refactor menu inner pointers

ppizarror opened the pull request.

anxuae commented on pygameMenu/menu.py:
```diff
@@ -1175,42 +1159,38 @@ def _handle_joy_event(self):
+        assert isinstance(events, (list, type(None)))
+
         break_mainloop = False
         if events is None:
```
the event should be get in the `mainloop` function
it shall not be optional in `update` one (as for widgets)

ppizarror replied:
true!

anxuae commented on pygameMenu/menu.py:
```diff
@@ -1332,26 +1305,38 @@ def _check_menu_initialized(self):
         return True
 
-    def mainloop(self, events=None, disable_loop=False):
+    def mainloop(self, surface, bgfun, events=None, disable_loop=False):
```
`events` parameter can be removed , no?

ppizarror replied:
yes

anxuae commented on pygameMenu/menu.py:
```diff
@@ -1332,26 +1305,38 @@ def _check_menu_initialized(self):
+            if self._top._enabled:
+                self._actual.draw(surface=surface)
+            _pygame.display.flip()
+            if break_mainloop or disable_loop:
```
check `self.is_disable` ?

ppizarror replied:
it's done before the while true

anxuae commented on pygameMenu/menu.py:
```diff
@@ -1332,26 +1305,38 @@ def _check_menu_initialized(self):
-        else:
-            self._main(events)
+        while True:
+            bgfun()
```
` bgfun()` should be called in the `draw()`method as it make a drawing ;-)

ppizarror replied:
> ` bgfun()` should be called in the `draw()`method as it make a drawing ;-)

but draw method does not require bgfun as a parameter. Should we add that?

anxuae replied:
I tank you can keep the class attribute `self._background_function`. Set it in the `mainloop` method and use it in the `draw` method.

This permit to keep a consistence in actions performed by each method.

ppizarror replied:
That's better

ppizarror commented:
@anxuae i think FPS should only be required by mainloop

anxuae commented:
> @anxuae i think FPS should only be required by mainloop

agree

ppizarror commented:
@anxuae added the background function to menu. Now mainloop does this:

```
def mainloop(self, surface, bgfun, disable_loop=False, fps_limit=0):
	"""
	Main function of menu.

	:param surface: Pygame surface to draw the menu
	:type surface: pygame.surface.SurfaceType
	:param bgfun: Background function called on each loop iteration before drawing the menu
	:type bgfun: function
	:param disable_loop: Disable infinite loop waiting for events
	:type disable_loop: bool
	:param fps_limit: Limit frame per second of the loop, if 0 there's no limit
	:type fps_limit: int,float
	:return: None
	"""
	assert isinstance(surface, _pygame.Surface)
	assert callable(bgfun), 'background function must be callable (a function)'
	assert isinstance(disable_loop, bool)
	assert isinstance(fps_limit, (int, float))
	assert fps_limit >= 0, 'fps limit cannot be negative'

	# Store the reference of the menu (if user moves through submenus)
	self._top = self

	if not self._top.is_enabled():
		return
	self._actual._background_function = bgfun
	while True:
		self._actual._clock.tick(fps_limit)
		break_mainloop = self.update(events=_pygame.event.get())
		if self._top.is_enabled():  # As event can change the status of the menu, this has to be checked twice
			self._actual.draw(surface=surface)
		_pygame.display.flip()
		if break_mainloop or disable_loop:
			self._actual._background_function = None
			return
```

ppizarror commented:
I added an exception if user tries to update the menu if it's not enabled

anxuae commented:
> @anxuae added the background function to menu. Now mainloop does this:

looks good to me

anxuae commented:
The variable `break_mainloop` in `update()` should be renamed to `updated` to keep same philosophy as in widgets.
In fact, Im not sure that you have to verify anymore the returned value of the `update()` in the `mainloop()`

ppizarror commented:
@anxuae with this PR timer_clock does not work, i mean as mainloop does not brake execution there's no way that a menu can't ignore the mainloop. Maybe a parameter `enable_mainloop` can solve this, if True a certain menu will not write any background image, and also while True would break at the first iteration

anxuae commented:
> @anxuae with this PR timer_clock does not work, i mean as mainloop does not brake execution there's no way that a menu can't ignore the mainloop. Maybe a parameter `enable_mainloop` can solve this, if True a certain menu will not write any background image, and also while True would break at the first iteration

I think that typically, the `timer_clock` should be the case 2 of #134 . The user want to do something during the loop. Thus the `mainloop()` method should not be used in this case.

Something like that should work:

```python

main_menu = Menu(...)

While True:

    events = pygame.event.get()
    for event in events:
        if event.type == pygame.QUIT:
            exit()

    # Draw timer
    if main_menu.get_title(current=True) == 'Timer Menu':
        time_string = str(datetime.timedelta(seconds=int(timer[0])))
        time_blit = timer_font.render(time_string, 1, COLOR_WHITE)
        time_blit_size = time_blit.get_size()
        surface.blit(time_blit, (int(W_SIZE / 2 - time_blit_size[0] / 2), int(H_SIZE / 2 - time_blit_size[1] / 2)))
    else:
        surface.fill(COLOR_BACKGROUND)

    main_menu.update(events)

    main_menu.draw(surface)

    pygame.display.update()
```

anxuae commented on pygameMenu/menu.py:
```diff
@@ -1,1614 +1,1581 @@
+
+        # Process events, check title
+        if self._actual._menubar.update(events):
+            break_mainloop = True
```
Updating the menubar should not break the mainloop.
I think that the flag `break_mainloop ` should be renamed in `updated` (as in widgets)

In fact the `update()` don't known if loop shall be broken or not. If user action has closed the menu, it will checked in the `mainloop()`

ppizarror replied:
But what would mean updated?

ppizarror commented:
> > @anxuae with this PR timer_clock does not work, i mean as mainloop does not brake execution there's no way that a menu can't ignore the mainloop. Maybe a parameter `enable_mainloop` can solve this, if True a certain menu will not write any background image, and also while True would break at the first iteration
> 
> I think that typically, the `timer_clock` should be the case 2 of #134 . The user want to do something during the loop. Thus the `mainloop()` method should not be used in this case.
> (this avoid a different loop management depending on the menu)
> 
> Something like that should work:
> 
> ```python
> main_menu = Menu(...)
> 
> While True:
> 
>     events = pygame.event.get()
>     for event in events:
>         if event.type == pygame.QUIT:
>             exit()
> 
>     # Draw timer
>     if main_menu.get_title(current=True) == 'Timer Menu':
>         time_string = str(datetime.timedelta(seconds=int(timer[0])))
>         time_blit = timer_font.render(time_string, 1, COLOR_WHITE)
>         time_blit_size = time_blit.get_size()
>         surface.blit(time_blit, (int(W_SIZE / 2 - time_blit_size[0] / 2), int(H_SIZE / 2 - time_blit_size[1] / 2)))
>     else:
>         surface.fill(COLOR_BACKGROUND)
> 
>     main_menu.update(events)
> 
>     main_menu.draw(surface)
> 
>     pygame.display.update()
> ```

After quite some work I added that feature, also improved the difference between actual and current. Renamed all the _actual to _current. Also improved the docs, some methods update the current menu (like adding widgets), if not, added a current=True/False to each public operation

ppizarror commented:
This PR is way too large, I'm gonna merge. @anxuae Please if anything does not work in the way intended, or want to improve the names (like update) please let's work on another PR 😄
